src/wallet.rs

```diff
@@ -33,18 +33,29 @@ pub trait GetUtxos {
     async fn get_utxos(&self, address: Address) -> Result<Vec<(OutPoint, TxOut)>>;
 }
 
-fn salt_encryption_key(
-    salt: impl AsRef<[u8]>,
-    encryption_key: impl AsRef<[u8]>,
-) -> Result<[u8; 32]> {
-    let salted_encryption_key = {
-        let h = Hkdf::<Sha256>::new(Some(salt.as_ref()), encryption_key.as_ref());
-        let mut enc_key = [0u8; 32];
-        h.expand(b"ENCRYPTION_KEY", &mut enc_key)
-            .context("failed to derive encryption key")?;
-        enc_key
-    };
-    Ok(salted_encryption_key)
+/// Derive the encryption key from the wallet's password and a salt.
+///
+/// # Choice of salt
+///
+/// The salt of HKDF can be public or secret and while it can operate without a salt, it is better to pass a salt value [0].
+///
+/// # Choice of ikm
+///
+/// The user's password is our input key material. The stronger the password, the better the resulting encryption key.
+///
+/// # Choice of info
+///
+/// HKDF can operate without `info`, however, it is useful to "tag" the derived key with its usage.
+/// In our case, we use the encryption key to encrypt the secret key and as such, tag it with `b"ENCRYPTION_KEY"`.
+///
+/// [0]: https://tools.ietf.org/html/rfc5869#section-3.1
+fn derive_encryption_key(salt: impl AsRef<[u8]>, password: impl AsRef<[u8]>) -> Result<[u8; 32]> {
+    let h = Hkdf::<Sha256>::new(Some(salt.as_ref()), password.as_ref());
+    let mut enc_key = [0u8; 32];
+    h.expand(b"ENCRYPTION_KEY", &mut enc_key)
+        .context("failed to derive encryption key")?;
+
+    Ok(enc_key)
 }
```

Comment:
I find it easier to understand code if the functions are ordered by their importance and call hierarchy. In this case, `initialize_new` calls `derive_encryption_key`, meaning `initialize_new` should come first.

The reasoning is that without know how or where `derive_encryption_key` is used, having to read it doesn't make any sense.

If all the public API and most used functions of a module are the first thing in a file, building up a mental model on what this module does is much easier. With appropriate naming of functions, you can understand a function like `initialize_new` even though it calls a function you haven't seen the implementation of.

Think of it like a book. The ToC gives you an overview of what the book is about through its headings (the public API) without having to read the content (the private utility functions) first.